File: C/Algorithms/1329.c
#include <stdlib.h>

int cmpi(const void *x, const void *y) {
  int a = *(const int *)x;
  int b = *(const int *)y;

  return (b > a) - (b < a);
}

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume
 * caller calls free().
 */
int **diagonalSort(int **mat, int matSize, int *matColSize, int *returnSize,
                   int **returnColumnSizes) {
  int rows = matSize - 1, cols = matColSize[0] - 1,
      // max size of diag is == rows.
                              *diag = malloc(matSize * sizeof(*diag)),
      diagLen = 0;
  // start from bottom left corner.
  int r = rows, c = 0;
  while (r != 0 || c != cols) {
    // Grab diagonal.
    int i = r, j = c;
    while (i <= rows && j <= cols)
      diag[diagLen++] = mat[i++][j++];

    // sort until diagLen.
    if (diagLen > 1)
      qsort(diag, diagLen, sizeof(*diag), cmpi);

    // Place back.
    i = r, j = c;
    while (i <= rows && j <= cols)
      mat[i++][j++] = diag[--diagLen];

    // Increment r, c
    if (r == 0)
      c++;
    if (r > 0)
      r--;
  }
  // Free and resign return values.
  free(diag);
  *returnColumnSizes = matColSize;
  *returnSize = matSize;
  return mat;
}

File: Rust/Algorithms/1329.rs
impl Solution {
    pub fn diagonal_sort(mut mat: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let (rows, cols) = (mat.len() - 1, mat[0].len() - 1);
        let mut diag = vec![];

        let (mut r, mut c) = (rows, 0);
        while r != 0 || c != cols {
            // Grab diagonal in diag.
            let (mut i, mut j) = (r, c);
            while i <= rows && j <= cols {
                diag.push(mat[i][j]);
                i += 1;
                j += 1;
            }
            // Sort it (reversed)
            diag.sort_unstable();
            diag = diag.into_iter().rev().collect();

            // And add it back to mat.
            i = r;
            j = c;
            while i <= rows && j <= cols {
                mat[i][j] = diag.pop().unwrap();
                i += 1;
                j += 1;
            }
            if r == 0 {
                c += 1;
            }
            if r > 0 {
                r -= 1;
            }
        }
        mat
    }
}
